File: frontend/types/industry-notification.ts
// Mirrors backend/app/schemas/industry_notification.py and
// database/migrations/055_industry_notifications.sql. Keep in sync.
//
// Notifications are produced by trusted system-context code (there is no
// insert policy on the table). The frontend only lists them, marks them
// read/unread, and reads the unread count. It never creates one.

export const NOTIFICATION_TYPES = ["NEW_APPLICATION", "WITHDRAWAL", "SYSTEM"] as const;
export type NotificationType = (typeof NOTIFICATION_TYPES)[number];

export const RELATED_ENTITY_TYPES = [
  "INTERNSHIP_APPLICATION",
  "JOB_APPLICATION",
  "PROJECT_APPLICATION",
  "WORKSHOP_APPLICATION",
  "TRAINING_APPLICATION",
] as const;
export type RelatedEntityType = (typeof RELATED_ENTITY_TYPES)[number];

export interface IndustryNotification {
  id: string;
  type: NotificationType;
  title: string;
  body: string;
  related_entity_type: RelatedEntityType | null;
  related_entity_id: string | null;
  is_read: boolean;
  read_at: string | null;
  created_at: string | null;
}

export interface IndustryNotificationListResponse {
  notifications: IndustryNotification[];
  unread_count: number;
}

/**
 * Maps a notification's related entity to an Industry route. For
 * INTERNSHIP_APPLICATION/JOB_APPLICATION, `related_entity_id` is the
 * application id (Industry already has a per-application detail route).
 * For PROJECT_APPLICATION/WORKSHOP_APPLICATION it is the POSTING id --
 * the destination is that posting's Applicants view, not a specific
 * applicant (see backend/app/services/notification_producer.py).
 */
export function relatedHref(n: IndustryNotification): string | null {
  const id = n.related_entity_id;
  if (!n.related_entity_type || !id) return null;
  switch (n.related_entity_type) {
    case "INTERNSHIP_APPLICATION":
    case "JOB_APPLICATION":
      return `/industry/applicants/${encodeURIComponent(id)}`;
    case "PROJECT_APPLICATION":
      return `/industry/projects/${encodeURIComponent(id)}/applicants`;
    case "WORKSHOP_APPLICATION":
      return `/industry/workshops/${encodeURIComponent(id)}/applicants`;
    case "TRAINING_APPLICATION":
      return `/industry/training/${encodeURIComponent(id)}/applicants`;
    default:
      return null;
  }
}

File: backend/app/schemas/industry_notification.py
"""Pydantic schemas for the INDUSTRY side of in-app notifications.

Field names and constraints match
database/migrations/055_industry_notifications.sql (`industry_notifications`)
exactly -- the Industry mirror of
backend/app/schemas/student_notification.py. There is deliberately no
create/write schema exposed to Industry: the migration has NO insert
policy, so notifications are only ever written by trusted system-context
code (service role, via app.services.notification_producer). The only
mutation an Industry account can make is toggling `read_at` on their own
row.
"""

from typing import Literal

from pydantic import BaseModel

# industry_notifications.type CHECK -- 055
NotificationType = Literal["NEW_APPLICATION", "WITHDRAWAL", "SYSTEM"]

# industry_notifications.related_entity_type CHECK -- 055. Always points
# at the POSTING (or, for INTERNSHIP/JOB, the application row -- Industry
# already has a per-application detail route) the event concerns.
RelatedEntityType = Literal[
    "INTERNSHIP_APPLICATION",
    "JOB_APPLICATION",
    "PROJECT_APPLICATION",
    "WORKSHOP_APPLICATION",
]


class IndustryNotification(BaseModel):
    """One `industry_notifications` row, the recipient's own. `is_read` is
    derived from `read_at`."""

    id: str
    type: str
    title: str
    body: str
    related_entity_type: str | None = None
    related_entity_id: str | None = None
    is_read: bool
    read_at: str | None = None
    created_at: str | None = None


class IndustryNotificationListResponse(BaseModel):
    notifications: list[IndustryNotification]
    unread_count: int


class MarkAllReadResponse(BaseModel):
    updated: int
